raft: remove commit index advancement dependence on heartbeats

iskettaneh opened the pull request:
This commit removes the commit index advancement's dependence on
heartbeats. It does so by making sure that the leader sends an empty
MsgApp request if it knows that the follower's commit index is stale.

Backward compatability arguments:

1) Followers can accept commit index advancement using both paths:
Heartbeats & MsgApps.

2) If the leader's point of view of the cluster version is old, it
uses the old path of advancing the followrs' commit index using
heartbeats.

3) If the leader's point of view of the cluster version is new, it
means that all nodes in the cluster (including all followers) have the
latest binary for sure. At that moment, the leader uses the new logic
of advancing the commit index using MsgApp rather than heartbeats.

Fixes: https://github.com/cockroachdb/cockroach/issues/125266

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129490)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 11 of 11 files at r1, 6 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129490)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @pav-kv)

___
*[`-- commits` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5F_Hq_1TWLb6f5Mzrf):*
"followers'"
___
*[`pkg/raft/raft.go` line 606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5F__U2DwdnqWLmSlZB:-O5F__U2DwdnqWLmSlZC:byvps6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/raft.go#L606)):*
> ```Go
> 
> 	last, commit := r.raftLog.lastIndex(), r.raftLog.committed
> 	advanceCommit := r.crdbVersion.IsActive(context.Background(),
> ```

Let's extract this version check into a well-named method on `raft` which we can call from each of the places where we need it.
___
*[`pkg/raft/raft.go` line 606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FcVYO16t-GfJq7n0j:-O5FcVYO16t-GfJq7n0k:bly2dmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/raft.go#L606)):*
> ```Go
> 
> 	last, commit := r.raftLog.lastIndex(), r.raftLog.committed
> 	advanceCommit := r.crdbVersion.IsActive(context.Background(),
> ```

It would be good to explain somewhere (maybe on the comment of the method I suggested above) what logic on the followers we're waiting for before starting to advance the commit index on MsgApps. In other words, why do we need a version gate?
___
*[`pkg/raft/raft.go` line 696 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5F_lIL8SYK3NmgSyrL:-O5F_lIL8SYK3NmgSyrM:b-epggf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/raft.go#L696)):*
> ```Go
> 		Match:  pr.Match,
> 	})
> 	if !r.crdbVersion.IsActive(context.Background(),
> ```

Instead of checking the version again (which is racy), we should have this condition be `if commit != 0 {`.
___
*[`pkg/raft/tracker/progress.go` line 318 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FcATyDOHrJhC_OfI6:-O5FcATyDOHrJhC_OfI7:b-jjkqr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L318)):*
> ```Go
> // In StateSnapshot, we do not send append messages.
> //
> // if advanceCommit is true, it means that MsgApp own the responsibility of
> ```

nit: "If".
___
*[`pkg/raft/tracker/progress.go` line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FcDH45-okPSc_KJwz:-O5FcDH45-okPSc_KJx-:b-t7pqr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L320)):*
> ```Go
> // if advanceCommit is true, it means that MsgApp own the responsibility of
> // making sure to advance the followers' commit index. If it's false, it means
> // that the responsibility is on heartbeats.
> ```

`s/heartbeats/MsgHeartbeat/` since you said MsgApp above.
___
*[`pkg/raft/tracker/progress.go` line 350 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FbxW-AtrNFTH-ksmO:-O5FbxW-AtrNFTH-ksmP:b-g1uxor) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L350)):*
> ```Go
> 		}
> 
> 		// Send an empty MsgApp containing the latest commit index if we know that
> ```

Can this case subsume the previous one? Or is there a reason why they need to remain separate? If so, it would be good to explain why we need both in the comments.

@pav-kv was recently in this code, so he should take a look at this as well.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129490)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @nvanbenschoten)

___
*[`pkg/raft/tracker/progress.go` line 350 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FbxW-AtrNFTH-ksmO:-O5Ow-m83NJ7BqQD54Li:b-xsf0lr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L350)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can this case subsume the previous one? Or is there a reason why they need to remain separate? If so, it would be good to explain why we need both in the comments.

@pav-kv was recently in this code, so he should take a look at this as well.
</blockquote></details>

I wonder if we can leave `ShouldSendMsgApp` almost intact (and not pass the `advanceCommit` flag) as follows:

- Merge `pr.sendCommit` and `receivedCommit` into something more generic like `minCommit`. The logic will be that `CanBumpCommit` checks `commit > minCommit`, and send a message if true, and we also would't need a second `IsFollowerCommitStale` method.
- Then the only difference on the leader is how to update the `minCommit`. At all times, bump `minCommit` from the `MsgAppResp` message. But if `advanceCommitViaMsgAppOnly` then additionally bump `minCommit` "optimistically" when sending `MsgApp`.

So it looks like it's possible to lift this behaviour difference out of `Progress` struct?
___
*[`pkg/raft/tracker/progress.go` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5OufHz5XCP0qzQwBte:-O5OufHz5XCP0qzQwBtf:byu48va) ([raw file](https://github.com/cockroachdb/cockroach/blob/466eb535fc645a21365481a83a783bab9b93578a/pkg/raft/tracker/progress.go#L59)):*
> ```Go
> 	sentCommit uint64
> 
> 	// receivedCommit is the highest commit index received from the follower.
> ```

Comment that the commit index is durable on the follower.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129490)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @nvanbenschoten)

___
*[`pkg/raft/tracker/progress.go` line 350 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FbxW-AtrNFTH-ksmO:-O5Oy7iV9gvD3Onsen-_:bf51ddh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L350)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I wonder if we can leave `ShouldSendMsgApp` almost intact (and not pass the `advanceCommit` flag) as follows:

- Merge `pr.sendCommit` and `receivedCommit` into something more generic like `minCommit`. The logic will be that `CanBumpCommit` checks `commit > minCommit`, and send a message if true, and we also would't need a second `IsFollowerCommitStale` method.
- Then the only difference on the leader is how to update the `minCommit`. At all times, bump `minCommit` from the `MsgAppResp` message. But if `advanceCommitViaMsgAppOnly` then additionally bump `minCommit` "optimistically" when sending `MsgApp`.

So it looks like it's possible to lift this behaviour difference out of `Progress` struct?
</blockquote></details>

There might be a slight difference in behaviours:

- with the previous logic, we wouldn't pace empty `MsgApp`s if we see that the commit index can be bumped. We would send them reactively as soon as `sentCommit < commit < Next`.
- with the new logic, the `receivedCommit`/`minCommit` wouldn't be bumped immediately, so we can spin if we do the same. So we should pace these messages in `receivedCommit < commit` case.

Maybe a combination of these two behaviours we want is:
- If `commit > receivedCommit`, consider sending a message.
- But if `sentCommit >= Next`, pace these messages.

Is this in the ballpark what the code does?


<!-- Sent from Reviewable.io -->

iskettaneh commented:
*[`pkg/raft/tracker/progress.go` line 350 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5FbxW-AtrNFTH-ksmO:-O5JD_jTDMTt2mnDX7TR:bcske80) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dad9d2e0cd46d653715e44b8c5a7d05ebe6d051/pkg/raft/tracker/progress.go#L350)):*
I didn't combine them into one condition because the two condition are slightly different:
1) The SentCommit as Pavel mentioned is not paced. Meaning that we can send two back-to-back messages if we know that the SentCommit is not to it's correct value. We can keep sending empty MsgApp until we advance the SentCommit all the way.
2) The new condition is paced. we don't want to keep sending empty MsgApp if we haven't received a MsgAppResp from a follower. That message might take some time until the leader receives it, and therefore it needs to be paced.


Regarding 

> But if `advanceCommitViaMsgAppOnly` then additionally bump `minCommit` "optimistically" when sending `MsgApp`.

I don't think we can advance the MinCommit in that case because the MsgApp might get dropped, and then if we advance the MinCommit, when will the leader send another MsgApp?


<!-- Sent from Reviewable.io -->

iskettaneh committed:
raft: remove commit index advancement dependence on heartbeats

This commit removes the commit index advancement's dependence on
heartbeats. It does so by making sure that the leader sends an empty
MsgApp request if it knows that the follower's commit index is stale.

Backward compatability arguments:

1) Followers can accept commit index advancement using both paths:
Heartbeats & MsgApps.

2) If the leader's point of view of the cluster version is old, it
uses the old path of advancing the followers' commit index using
heartbeats.

3) If the leader's point of view of the cluster version is new, it
means that all nodes in the cluster (including all followers) have the
latest binary for sure. At that moment, the leader uses the new logic
of advancing the commit index using MsgApp rather than heartbeats.

Fixes: #125266

Epic: None

Release note: None

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -675,14 +676,23 @@ func (r *raft) sendHeartbeat(to pb.PeerID) {
-	commit := min(pr.Match, r.raftLog.committed)
+	// NOTE: Starting from V24_3_AdvanceCommitIndexViaMsgApps, heartbeats do not
+	// advance the commit index. Instead, MsgApp are used for that purpose.
+	// TODO(iskettaneh): Remove the commit from the heartbeat message once no
```
Consider expanding "no longer needed", e.g. saying that the `Commit` code here can be removed in version >= 25.1.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -1997,9 +2007,12 @@ func (r *raft) handleHeartbeat(m pb.Message) {
-	mark := LogMark{Term: m.Term, Index: min(m.Commit, r.raftLog.lastIndex())}
-	if mark.Term == r.raftLog.accTerm() {
-		r.raftLog.commitTo(mark)
+	// TODO(iskettaneh): Remove this logic once no longer needed.
+	if m.Commit != 0 {
+		mark := LogMark{Term: m.Term, Index: min(m.Commit, r.raftLog.lastIndex())}
+		if mark.Term == r.raftLog.accTerm() {
+			r.raftLog.commitTo(mark)
+		}
 	}
```
Probably don't need to change this bit and the TODO, because `commitTo` is a no-op if `Commit` <= the current commit index, which is trivially true when `Commit == 0`.

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -2324,3 +2337,12 @@ func (r *raft) testingStepDown() error {
 	return nil
 }
+
+// onlyAdvanceCommitViaMsgApp returns true if the commit index is advanced on
```
nit: fix the name, it should match the func name

pav-kv commented on pkg/raft/raft.go:
```diff
@@ -2324,3 +2337,12 @@ func (r *raft) testingStepDown() error {
+// the followers using MsgApp only. This means that heartbeats are not used to
+// advance the commit index. This function returns true only if followers
+// understand that heartbeats might not contain a commit index.
+func (r *raft) advanceCommitViaMsgAppOnly() bool {
```
nit: I think the precondition (that we're gating against) is not `if followers understand that heartbeats might not contain a commit index`, but rather: `if all followers communicate their durable commit index back to the leader via MsgAppResp`.

How about something like:
```
advanceCommitViaMsgAppOnly returns true if the leader can stop communicating
the commit index to followers using MsgHeartbeat, and instead do it reliably
via MsgApp. This is safe when all followers send their durable commit index
back to the leader via MsgAppResp.
```

pav-kv commented on pkg/raft/testdata/msg_app_commit_index.txt:
```diff
@@ -0,0 +1,191 @@
+
+# The network blip ends here.
+
+status 1
```
It could be good to add `sentCommit` and `matchCommit` to status here. It would change a lot of tests though, consider doing in a follow-up.

pav-kv commented on pkg/raft/testdata/msg_app_commit_index.txt:
```diff
@@ -0,0 +1,191 @@
+
+# On the next MsgApp sent to follower 3, the leader will include that the
+# commit index is 13. Notice that the leader doesn't send MsgApp to follower 2
+# because it knows that it has the latest commit index.
```
Nice test.

pav-kv commented on pkg/raft/testdata/msg_app_commit_index_leader_old_version.txt:
```diff
@@ -0,0 +1,154 @@
+# This test demonstrates the case where the new commit index advancement logic
+# via MsgApp landed on some nodes, but not the leader. In this case,
+# heartbeats will continue advancing the commit index normally.
```
TODO to remove this test when the code has fully migrated to the new behaviour?

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -56,6 +56,10 @@ type Progress struct {
 	// In StateSnapshot, sentCommit == PendingSnapshot == Next-1.
 	sentCommit uint64
 
+	// matchCommit is the commit at which the follower is known to match the
```
nit: `is the commit index at which ...`

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -295,7 +315,11 @@ func (pr *Progress) IsPaused() bool {
+//
+// If advanceCommit is true, it means that MsgApp owns the responsibility of
+// making sure to advance the followers' commit index. If it's false, it means
+// that the responsibility is primarily on MsgHeartbeat.
```
nit: the responsibility is still "primarily" on MsgApp, the heartbeats only close the gap in reliability in cases when MsgApp are dropped

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -320,7 +344,26 @@ func (pr *Progress) ShouldSendMsgApp(last, commit uint64) bool {
+		// send this message if pr.MsgAppProbesPaused is false. After this message,
+		// pr.MsgAppProbesPaused will be set to true until we receive a heartbeat
+		// response from the follower. In contrast, the condition above can keep
+		// sending empty MsgApps until we have sent the latest commit index to the
```
`sending empty MsgApps eagerly`

pav-kv reviewed (approved):
LGTM % a few nits

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -196,6 +200,14 @@ func (pr *Progress) CanBumpCommit(index uint64) bool {
 
+// IsFollowerCommitStale returns true if the follower's commit index it less
+// than index.
+// If the follower's commit index is pr.Next-1, it means that sending a larger
```
This comment seems to be not true? The `pr.Next` logic sits in `CanBumpCommit`. Did you intend to check it similarly here?

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -320,7 +344,26 @@ func (pr *Progress) ShouldSendMsgApp(last, commit uint64) bool {
+		// TODO(iskettaneh): Remove the dependency on MsgAppProbesPaused to send
+		// MsgApps.
+		if advanceCommit {
+			return pr.IsFollowerCommitStale(commit) && !pr.MsgAppProbesPaused
```
Consider inlining `IsFollowerCommitStale`, analogous to `if pr.Match < last && !paused` check above. Or don't do it if you intend to make the condition in it more complex (like check `pr.Next` too, see the corresponding comment above).

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129490)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, @nvanbenschoten, and @pav-kv)

___
*[`pkg/raft/raft.go` line 681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738359206:-O5YhDCsBQF9O1-ksVBs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/raft.go#L681)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Consider expanding "no longer needed", e.g. saying that the `Commit` code here can be removed in version >= 25.1.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 2016 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738368053:-O5YhY8U8Gv1Uei0CzBd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/raft.go#L2016)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Probably don't need to change this bit and the TODO, because `commitTo` is a no-op if `Commit` <= the current commit index, which is trivially true when `Commit == 0`.
</blockquote></details>

Done.
___
*[`pkg/raft/tracker/progress.go` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129490#-O5OufHz5XCP0qzQwBte:-O5U8KL396Io9ZKTbDjw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/466eb535fc645a21365481a83a783bab9b93578a/pkg/raft/tracker/progress.go#L59)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Comment that the commit index is durable on the follower.
</blockquote></details>

Done.
___
*[`pkg/raft/tracker/progress.go` line 205 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738409186:-O5Ykphx0uYNT3GntRTt:byfpu0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/tracker/progress.go#L205)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This comment seems to be not true? The `pr.Next` logic sits in `CanBumpCommit`. Did you intend to check it similarly here?
</blockquote></details>

Yes! I had the idea of doing something very similar.
___
*[`pkg/raft/tracker/progress.go` line 359 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738403708:-O5YlmAU0ePraeBk89Mx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/tracker/progress.go#L359)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`sending empty MsgApps eagerly`
</blockquote></details>

Done.
___
*[`pkg/raft/tracker/progress.go` line 364 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738411203:-O5YlpOV-Az_ei6h-tIA:byn7dpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/tracker/progress.go#L364)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Consider inlining `IsFollowerCommitStale`, analogous to `if pr.Match < last && !paused` check above. Or don't do it if you intend to make the condition in it more complex (like check `pr.Next` too, see the corresponding comment above).
</blockquote></details>

Yeah, I decided to make the condition above a bit more complicated. So I will keep it in a separate function
___
*[`pkg/raft/testdata/msg_app_commit_index.txt` line 110 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738393252:-O5Z-Ec61rYqs3q8MqrM:bgzhpwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/testdata/msg_app_commit_index.txt#L110)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

It could be good to add `sentCommit` and `matchCommit` to status here. It would change a lot of tests though, consider doing in a follow-up.
</blockquote></details>

Will do it in a follow-up pr
___
*[`pkg/raft/testdata/msg_app_commit_index.txt` line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129490#gh-1738395141:-O5YgQkp9rSKw1FzXnOh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/222f1a6d1a05717b630a1f9cb4c05c1a81b3f972/pkg/raft/testdata/msg_app_commit_index.txt#L130)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Nice test.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

iskettaneh commented:
bors r+
